src/meta/expr.lean

```diff
@@ -286,6 +309,30 @@ meta def instantiate_lambdas_or_apps : list expr → expr → expr
 | (e'::es) (lam n bi t e) := instantiate_lambdas_or_apps es (e.instantiate_var e')
 | es       e              := mk_app e es
 
+/-- Get the codomain/target of a pi-type. Doesn't Instantiate bound variables. -/
+meta def pi_codomain : expr → expr
```

Comment:
This is redundant with `mk_local_pis`

Reply:
`get_pi_arity` is slower than just looking at the expression, defined in `pi_arity` (see docstring of `pi_arity`).
For the same reason I thought that `mk_local_pis` would be slower than what I implemented here. Unless `expr.instantiate_var` is `O(1)`, I think this is significantly quicker.
I now see that there is a difference between `get_pi_arity` and `mk_local_pis`: `get_pi_arity` performs `whnf`, so maybe that causes the slowdown.

Reply:
If you want  can add a comment to that effect to the docstring and/or I can also empirically test it.

Reply:
I see what you mean. What's your story with bound variables? Why is it ok to not instantiate them?

Reply:
It depends on what you want to do with them. I'm not doing anything that assumes that the resulting expressions are well-formed (no `infer_type` or `unify` or anything like that). I just do some syntactic checks on them (does it contain variable `i`? Are the variables occurring in different places equal?)

Reply:
I think it's a reasonable idea. What I'm struggling with is that most of the API otherwise strives to keep expressions closed. 

I can think of two ways of sticking to that convention: make those functions private or design a data structure to make substitution cheaper. 

I would go for the first one as the second one is a project in itself. But if this is the road you want to take here is what I'm considering for the data structure:

```lean
meta inductive expr2 : Type
| lift : expr -> expr2
| subst : expr2 -> nat -> expr -> expr2
```

That would let you stack up substitutions without applying until you're ready or not at all if you don't actually need it. We can also do partial substitution application 

```lean
meta inductive expr2 : Type
| lift : expr -> expr2
| app : expr2 -> expr2 -> expr2
| subst : expr2 -> nat -> expr -> expr2
-- ...
```

Reply:
How about just adding a warning that the resulting expressions are not closed?

Reply:
That could work. I think it would be good if it was specific and mention which functions (e.g. `infer_type` and `unify`) would break if called on those functions.

Also, can we put them in a namespace like `unsafe` to mark them as tricky?

Reply:
I think using an `unsafe` namespace is overkill. A warning in the doc string seems fine to me. If someone does misuse it, it shouldn't be hard to track down the problem.

Reply:
I disagree. I think a point could even be made to put `unsafe_` as part of the name itself. You're actually breaking important invariants here. It should be hard to miss.

Reply:
I made a note about open expressions, expanded the doc-strings. I added `unsafe_` to the name, although I don't really like that. Do you think it would be hard to miss if we dropped `unsafe_`?

Reply:
Sounds good. Good doc strings (which we now have in this case) are much better than umambiguous `unsafe_`s.

Reply:
The consensus seems to be to remove `unsafe` prefixes, which I did.